pennylane/ops/functions/assert_valid.py

```diff
@@ -0,0 +1,274 @@
+        _assert_error_raised(
+            op.diagonalizing_gates, qml.operation.DiagGatesUndefinedError, failure_comment
+        )()
+
+    has_eigvals = True
+    try:
+        eg = op.eigvals()
+        compute_eg = op.compute_eigvals(*op.data, **op.hyperparameters)
+        assert qml.math.allclose(eg, compute_eg), "eigvals and compute_eigvals must match"
+    except EigvalsUndefinedError:
+        has_eigvals = False
+
+    if has_eigvals and op.has_diagonalizing_gates:
+        dg = qml.prod(*dg) if len(dg) > 0 else qml.Identity(op.wires)
+        eg = qml.Hermitian(np.diag(eg), wires=op.wires)
+        decomp = qml.prod(qml.adjoint(dg), eg, dg)
+        decomp_mat = qml.matrix(decomp)
+        original_mat = qml.matrix(op)
+        failure_comment = f"eigenvalues and diagonalizing gates must be able to reproduce the original operator. Got \n{decomp_mat}\n\n{original_mat}"
+        assert qml.math.allclose(decomp_mat, original_mat), failure_comment
+
+
+def _check_copy(op):
+    """Check that copies and deep copies give identical objects."""
+    copied_op = copy.copy(op)
+    assert qml.equal(copied_op, op), "copied op must be equal with qml.equal"
+    assert copied_op == op, "copied op must be equivalent to original operation"
+    assert copied_op is not op, "copied op must be a separate instance from original operaiton"
+    assert qml.equal(copy.deepcopy(op), op), "deep copied op must also be equal"
+
+
+# pylint: disable=import-outside-toplevel, protected-access
+def _check_pytree(op):
+    """Check that the operator is property a pytree."""
+    data, metadata = op._flatten()
+    try:
+        assert hash(metadata), "metadata must be hashable"
+    except Exception as e:
+        raise ValueError(
+            f"metadata output from _flatten must be hashable. Got metadata {metadata}"
+        ) from e
+    new_op = type(op)._unflatten(data, metadata)
+    assert op == new_op, "metadata and data must be able to replicate the original operation"
+
+    try:
+        import jax
+    except ImportError:
+        return
+    leaves, struct = jax.tree_util.tree_flatten(op)
+    unflattened_op = jax.tree_util.tree_unflatten(struct, leaves)
+    assert unflattened_op == op, f"op must be a valid pytree. Got {unflattened_op} instead of {op}."
+
+    for d1, d2 in zip(op.data, leaves):
+        assert qml.math.allclose(
+            d1, d2
+        ), f"data must be the terminal leaves of the pytree. Got {d1}, {d2}"
+
+
+def _check_pickle(op):
+    """Check that an operation can be dumped and reloaded with pickle."""
+    pickled = pickle.dumps(op)
+    unpickled = pickle.loads(pickled)
+    assert unpickled == op, "operation must be able to be pickled and unpickled"
+
+
+# pylint: disable=no-member
+def _check_bind_new_parameters(op):
+    """Check that bind new parameters can create a new op with different data."""
+    new_data = [d * 0.0 for d in op.data]
+    new_data_op = qml.ops.functions.bind_new_parameters(op, new_data)
```

Comment:
Love this! It'll make it so much easier to find which templates might need custom dispatches for `bind_new_parameters` (although I already know the answer to that question is "most of them")

Reply:
I stand corrected. Just saw the tests, so many templates are working!